libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UIViewModel.kt

```diff
@@ -25,18 +23,27 @@ import kotlinx.coroutines.launch
 @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
 abstract class UIViewModel<State, Action>(initialState: State) : MapboxNavigationObserver {
 
-    private val _action = MutableSharedFlow<Action>(extraBufferCapacity = 1)
-    val action: Flow<Action> = _action
-    protected val _state: MutableStateFlow<State> = MutableStateFlow(initialState)
+    private val _state: MutableStateFlow<State> = MutableStateFlow(initialState)
     val state: StateFlow<State> = _state
 
     lateinit var mainJobControl: JobControl
 
     /**
      * Invoke an action for the behavior.
      */
-    fun invoke(action: Action) {
-        _action.tryEmit(action)
+    fun invoke(action: Action): State {
+        val mapboxNavigation: MapboxNavigation? = MapboxNavigationApp.current()
+        return if (mapboxNavigation != null) {
+            process(mapboxNavigation, _state.value, action).also {
+                _state.value = it
+            }
```

Comment:
```suggestion
            _state.value = process(mapboxNavigation, _state.value, action)
```

Reply:
yeah i wish too!
```
Assignments are not expressions, and only expressions are allowed in this context
```

![Screen Shot 2022-04-06 at 11 10 42 AM](https://user-images.githubusercontent.com/3021882/162040554-19711a42-09ab-4586-85e6-baa66059092d.png)

Reply:
Oh... you made `invoke` return `State`? Why?

Reply:
Yeah agreed that this looks a bit funky that we are returning `State` upon calling `invoke`
@kmadsen

Reply:
Removing race conditions when you need to invoke and get the state.

``` kotlin
val state = viewModel.invoke(..)
```
```
componentA.viewModel.invoke(..)
componentB.viewModel.invoke(..)
val state componentA.viewModel.state.value
```

```
val state = routesViewModel.invoke(RoutesAction.FetchPoints(listOf(lastPoint, destination.point)))
assertTrue(state is Fetching)
```